File: 399. Evaluate Division/main.cpp
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
    void dfs(unordered_map<string, vector<pair<string, double>>> &maps,
             unordered_set<string> &visited,
             string &src,
             string &dest,
             double value,
             double &answer) {
        visited.insert(src);

        if (src == dest) {
            answer = value;
            return;
        }

        for (auto &pair : maps[src]) {
            if (!visited.contains(pair.first)) {
                dfs(maps, visited, pair.first, dest, value * pair.second, answer);
            }
        }

        visited.erase(visited.find(src));
    }

public:
    vector<double> calcEquation(vector<vector<string>> &equations,
                                vector<double> &values,
                                vector<vector<string>> &queries) {
        unordered_map<string, vector<pair<string, double>>> maps;

        for (int i = 0; i < equations.size(); i++) {
            maps[equations[i][0]].push_back({equations[i][1], values[i]});
            maps[equations[i][1]].push_back({equations[i][0], 1.0 / values[i]});
        }

        vector<double> result;

        for (int i = 0; i < queries.size(); i++) {
            if (!maps.contains(queries[i][0]) || !maps.contains(queries[i][1])) {
                result.push_back(-1.0);
                continue;
            }

            unordered_set<string> visited;
            double answer = -1.0;
            dfs(maps, visited, queries[i][0], queries[i][1], 1.0, answer);

            result.push_back(answer);
        }

        return result;
    }
};

class Solution {
    unordered_map<string, pair<string, double>> maps;

    pair<string, double> find(string x) {
        if (maps.find(x) == maps.end()) {
            maps[x] = {x, 1.0};
        }

        auto [y, valueY] = maps[x];

        if (y != x) {
            auto [z, valueZ] = find(y);
            maps[x] = {z, valueY * valueZ};
        }

        return maps[x];
    }

    void unions(string u, string v, double value) {
        auto [rootX, valueX] = find(u);
        auto [rootY, valueY] = find(v);

        if (rootX != rootY) {
            maps[rootX] = {rootY, value * valueY / valueX};
        }
    }

public:
    vector<double> calcEquation(vector<vector<string>> &equations,
                                vector<double> &values,
                                vector<vector<string>> &queries) {
        size_t size = equations.size();

        for (size_t i = 0; i < size; i++) {
            unions(equations[i][0], equations[i][1], values[i]);
        }

        vector<double> result;

        for (auto &query : queries) {
            string u = query[0];
            string v = query[1];

            if (!maps.contains(u) || !maps.contains(v)) {
                result.push_back(-1.0);
                continue;
            }

            auto [rootU, valueU] = find(u);
            auto [rootV, valueV] = find(v);

            if (rootU != rootV) {
                result.push_back(-1.0);
            } else {
                result.push_back(valueU / valueV);
            }
        }

        return result;
    }
};

File: 399. Evaluate Division/main.py
from collections import defaultdict
from typing import Dict, List, Set, Tuple


class Solution:
    def calcEquation(
        self, equations: List[List[str]], values: List[float], queries: List[List[str]]
    ) -> List[float]:
        def dfs(
            maps: Dict[str, List[Tuple[str, float]]],
            visited: Set[str],
            src: str,
            dest: str,
            value: float,
        ):
            nonlocal answer

            visited.add(src)

            if src == dest:
                answer = value
                return

            for s, v in maps[src]:
                if s not in visited:
                    dfs(maps, visited, s, dest, value * v)

            visited.remove(src)

        maps = defaultdict(list)
        for i in range(len(equations)):
            maps[equations[i][0]].append((equations[i][1], values[i]))
            maps[equations[i][1]].append((equations[i][0], 1.0 / values[i]))

        result = list()

        for query in queries:
            if query[0] not in maps or query[1] not in maps:
                result.append(-1.0)
                continue

            visited = set()
            answer = -1.0

            dfs(maps, visited, query[0], query[1], 1.0)

            result.append(answer)

        return result


class Solution:
    def calcEquation(
        self, equations: List[List[str]], values: List[float], queries: List[List[str]]
    ) -> List[float]:
        def find(x: str) -> Tuple[str, float]:
            if x not in maps:
                maps[x] = (x, 1.0)

            (y, valueY) = maps[x]

            if y != x:
                (z, valueZ) = find(y)
                maps[x] = (z, valueY * valueZ)

            return maps[x]

        def union(u: str, v: str, val: float):
            (rootU, valueU) = find(u)
            (rootV, valueV) = find(v)

            if rootU != rootV:
                maps[rootU] = (rootV, val * valueV / valueU)

        maps = defaultdict(list)
        result = list()

        for u, v in queries:
            if u not in maps or v not in maps:
                result.append(-1.0)
                continue

            (rootU, valueU) = find(u)
            (rootV, valueV) = find(v)

            if rootU != rootV:
                result.append(-1.0)
            else:
                result.append(valueU / valueV)

        return result
